### PR TITLE
Add test cases for nullable items

### PR DESCRIPTION
**Pull request by MaryGao**
# Cadl Ranch Contribution Checklist:

- [x] I have written a [scenario spec](../docs/writing-scenario-spec.md)
- [x] I have **meaningful** `@scenario` names. Someone can look at the list of scenarios and understand what I'm covering.
- [x] I have written a [mock API](../docs/writing-mock-apis.md)
- [x] I have used `@scenarioDoc`s for extra scenario description and to tell people **how to pass** my mock api check.

### CODE REVIEW

**Review by pshao25 — CHANGES_REQUESTED** [review]
Can we remove the union part? It is not going to be a valid case in .Net and Java.

File: packages/cadl-ranch-specs/cadl-ranch-summary.md [hunk 1]
```diff
@@ -1413,6 +1413,26 @@ Expected input body:
...
+```
+
+### Models_Property_Types_NullableModel_put
+
+- Endpoint: `put /models/properties/types/nullable-model`
+
+Expected input body:
+
+```json
+{ "property": null }
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/cadl-ranch-summary.md:1433]: Personally, as REST API, I don't think PUT `{ "property": null }` semantically be any different from a PUT with `{}`.

But I won't block the case.

Comment (lirenhe) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1433]: Will JS treat '{}' as undefined and 'null' as null?

Comment (MaryGao) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1433]: Let me change the test case to JSON MERGE PATCH with partial update - https://www.rfc-editor.org/rfc/rfc7386#section-1.

```
{
         "a":"z",
         "c": {
       "f": null
         }
       }

```

**Comment by MaryGao** [comment]
> Can we remove the union part? It is not going to be a valid case in .Net and Java.

@pshao25 @weidongxu-microsoft Is this decision finalized that we don't support nullable model/array/dictionary in .Net or Java? Or this is just current status.

**Review by pshao25 — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
> > Can we remove the union part? It is not going to be a valid case in .Net and Java.
> 
> @pshao25 @weidongxu-microsoft Is this decision finalized that we don't support nullable model/array/dictionary in .Net or Java? Or this is just current status.

What is the "union part" about?

Java support nullable in array/dictionary, and nullable model for json merge patch.

File: packages/cadl-ranch-specs/cadl-ranch-summary.md (hunk 1 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1433]: Yes, json merge patch would be great!

File: packages/cadl-ranch-specs/cadl-ranch-summary.md [hunk 2]
```diff
@@ -1513,6 +1553,29 @@ Expected input body:
...
+
+### Models_Property_Types_ModelNullableProperty_patch
+
+- Endpoint: `patch /models/properties/types/nullable-property`
+
+Expected input body:
+
+```json
+{ "id": 1, "property": null }
+```
```
Comment (MaryGao) [inline packages/cadl-ranch-specs/cadl-ranch-summary.md:1577]: Here is the merge patch with partial update case.

File: packages/cadl-ranch-specs/http/models/property-types/main.tsp [hunk 3]
```diff
@@ -170,16 +170,78 @@ model CollectionsModelProperty is ModelTemplate<InnerModel[]>;
...
 @operationGroup
 interface DictionaryString extends ModelOperations<DictionaryStringProperty, "{'k1': 'hello', 'k2': 'world'}"> {}
 
+// Test a model with a nullable dictionary property
+@doc("Model with collection with nullable values")
+model DictionaryNullableValueProperty is ModelTemplate<Record<string | null>>;
+@route("/dictionary/nullable-value")
+@operationGroup
+interface DictionaryNullableValue
+  extends ModelOperations<DictionaryNullableValueProperty, "{'k1': 'hello', 'k2': 'world', 'k3': null}"> {}
```
Comment (iscai-msft) [inline packages/cadl-ranch-specs/http/models/property-types/main.tsp:193]: what happens if we instead just have a nullable dictionary / array? (no model involved). would be good to have tests for that as well

Comment (iscai-msft) [reply packages/cadl-ranch-specs/http/models/property-types/main.tsp:193]: oh i see @MaryGao already left a comment about this, i'll follow the response to that question

**Comment by iscai-msft** [comment]
> > Can we remove the union part? It is not going to be a valid case in .Net and Java.
> 
> @pshao25 @weidongxu-microsoft Is this decision finalized that we don't support nullable model/array/dictionary in .Net or Java? Or this is just current status.

i also have this question, thanks @MaryGao !

File: packages/cadl-ranch-specs/cadl-ranch-summary.md [hunk 4]
```diff
@@ -1513,6 +1553,29 @@ Expected input body:
 {"property": <don't include this property>}
 ```
 
+### Models_Property_Types_ModelNullableProperty_get
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: Do those test belong in the Model Properties scenario and not their coresponding container
- Collection
- Dictionary

**Comment by weidongxu-microsoft** [comment]
> > > Can we remove the union part? It is not going to be a valid case in .Net and Java.
> > 
> > 
> > @pshao25 @weidongxu-microsoft Is this decision finalized that we don't support nullable model/array/dictionary in .Net or Java? Or this is just current status.
> 
> i also have this question, thanks @MaryGao !

Already replied in https://github.com/Azure/cadl-ranch/pull/234#issuecomment-1469059819
No such decision in Java (no idea why the idea).

File: packages/cadl-ranch-specs/http/models/property-types/main.tsp [hunk 5]
```diff
@@ -170,16 +170,78 @@ model CollectionsModelProperty is ModelTemplate<InnerModel[]>;
...
+@operationGroup
+interface ModelNullableProperty {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "For testing"
+  @scenario
+  @scenarioDoc("""
+Expected response body::
+  ```json
+  {
+    "id": 1,
+    "property": "foo"
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/models/property-types/main.tsp:223]: It is `ModelNullableProperty` test, do you want property="foo" or property=null?

Comment (MaryGao) [reply packages/cadl-ranch-specs/http/models/property-types/main.tsp:223]: I design the case the property could be string or null, so here in `get` method it would be a foo string, and in `patch` method it would be a explicit null value.

File: .changeset/real-pets-mix.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"@azure-tools/cadl-ranch-specs": minor
+---
+
+Add test cases for nullable values
```
[the remark this answers is not in the document]
Comment (MaryGao) [reply .changeset/real-pets-mix.md:5]: Could you explore when it should be a patch and when should be a minor?

Here I add new subroute under property-types so I think it's minor change.

File: packages/cadl-ranch-specs/cadl-ranch-summary.md (hunk 4 above)
Comment (MaryGao) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: I will keep the current position for these cases. Because the collection and dictionary cases are focusing on testing the response/request body as arrays or records. 

However in our nullable vaules we return normal models in which we contains some properties with nullable values.

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: Maybe I am misunderstanding but from what I see is the array iteself contains nullable value and the dictionary itself contains nullable values not the property that is of type array/dictionary being null which is what this spec is meant to test.

Comment (MaryGao) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: Got your point. I think we more likely request/response a payload in which its has a array property and in this container it may cover nullable item. I think in reality the more meaningful case would be the latter one. You could see the issue https://github.com/Azure/autorest.csharp/issues/3084.

```
alias nullableFloat = float32 | null;

model MyModel {
  listOfFloats?: nullableFloat[];   
}
```

This is why I design this case as property not array or dictionary. But we could add these cases in both position. How do you think?

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: I totally understand that this is most likely what you’ll see in services but I wanted to keep things to the minimum required in Cadl ranch and in this case the mode property  bring no value.

op getNullableArray(): nullalbleFloat[] is i feel like what we really care to test here.

Comment (timotheeguerin) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: Unless I’m missing something and there is something actually different about it being in a property

Comment (MaryGao) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: Okay thanks for the explaination! I think it makes sense and I'll change them to collection/dictionary sections.

Comment (MaryGao) [reply packages/cadl-ranch-specs/cadl-ranch-summary.md:1556]: I adjust the cases:

- Arrays/nullable-float -> NullableFloat[]
- Dictionary/nullable-float -> Record<NullableFloat>
- PropertyTypes/nullable-property -> ModelWithNullableProperty

File: packages/cadl-ranch-specs/http/models/property-types/main.tsp [hunk 7]
```diff
@@ -183,3 +183,50 @@ model NeverProperty is ModelTemplate<never>;
...
 interface Never extends ModelOperations<NeverProperty, "<don't include this property>"> {}
+
+// Test a model with a model property which could be nullable
+@doc("Model with nullable properties")
+model ModelWithNullableProperty {
+  @doc("id property")
+  id: string;
+
+  @doc("nullable property")
+  property: string | null;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/property-types/main.tsp:194]: Sorry to bring new things but do we only care to test string or should we have a larger coverage of types that can be `null` as property.
i.e. Should we have a dedicated nullable model property spec that is pretty much like the optional model property spec but with `| null` instead of `?:`

File: packages/cadl-ranch-specs/http/models/property-types/main.tsp [hunk 8]
```diff
@@ -183,3 +183,50 @@ model NeverProperty is ModelTemplate<never>;
...
+  ```json
+  {
+    "id": 1,
+    "property": "foo"
+  }
+  ```
+""")
+  @get
+  @doc("Get call")
+  get(): ModelWithNullableProperty;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/property-types/main.tsp:214]: Should we maybe not test the case where the service returns `null` instead(Or maybe both)

Comment (MaryGao) [reply packages/cadl-ranch-specs/http/models/property-types/main.tsp]: Similar optional property spec I keep both the getAll and getDefault. The first would return non-null value the latter would be a null.

File: .changeset/real-pets-mix.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaryGao) [reply .changeset/real-pets-mix.md]: @timotheeguerin Do you think this is a patch or minor?

[the remark this answers is not in the document]
Comment (timotheeguerin) [reply .changeset/real-pets-mix.md]: Don't have a super strong opinion but until we are in v1 we can follow the following rule:
- patch: new spec or fix
- minor: If there is something that actually change the flow(Breaking changes)
- NO Major
After release of dpg and we have a good set of spec we can move to semantic versioning so any new spec will increase minor

File: packages/cadl-ranch-specs/http/models/property-types/main.tsp (hunk 8 above)
Comment (timotheeguerin) [reply packages/cadl-ranch-specs/http/models/property-types/main.tsp]: awesome!

File: packages/cadl-ranch-specs/http/models/property-nullable/mockapi.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Return the get and put operations
+ * @param route The route within /models/properties/optional/all/ for your function.
+ * @param value The value you are expecting and will return
+ */
+function createMockApis(route: string, value: any): MockApiGetPut {
+  const url = `/models/properties/nullable/${route}`;
+  const allUrl = `${url}/all`;
+  const defaultUrl = `${url}/default`;
```
Comment (timotheeguerin) [inline packages/cadl-ranch-specs/http/models/property-nullable/mockapi.ts:20]: Would just say here that `default` doesn't make that much sense for nullable as null is a value and we are not involving the default values.

Maybe name `withNull` and `withoutNull` or anything else you feel is more descriptive of the fact that we are testing sending null or not.

**Review by timotheeguerin — APPROVED** [review]
Looks great! made a last comment on the naming of the default operation, approving feel free to merge after resolving that!
